execute/observation.go

```diff
@@ -259,68 +261,158 @@ func buildCombinedReports(
+				totalMsgs++
+			} else {
+				// This is for when we calculate roots in reports later, we need the seqNum and
+				// the hash for this message
+				messageObs[srcChain][seqNum] = createEmptyMessageWithSeqNum(seqNum)
+			}
+
+			// Compute hash for all messages in report even if they are executed or not included in this round
+			hash, err := p.msgHasher.Hash(ctx, msg)
 			if err != nil {
-				lggr.Errorw("unable to read all messages for report",
-					"srcChain", srcChain,
-					"seqRange", report.SequenceNumberRange,
-					"merkleRoot", report.MerkleRoot,
-					"err", err,
+				return exectypes.Observation{}, fmt.Errorf(
+					"unable to hash message srcChain %d, seqNum %d, msgId %v: %w",
+					srcChain, seqNum, msg.Header.MessageID, err,
 				)
-				continue
 			}
+			msgHashes[srcChain][seqNum] = hash
 
-			if !msgsConformToSeqRange(msgs, report.SequenceNumberRange) {
-				lggr.Errorw("missing messages in range",
-					"srcChain", srcChain, "seqRange", report.SequenceNumberRange)
-				continue
-			}
+			// Update the observation with current state
+			observation.CommitReports = availableReports
+			observation.Messages = messageObs
+			observation.Hashes = msgHashes
 
-			for _, msg := range msgs {
-				messageObs[srcChain][msg.Header.SequenceNumber] = msg
-				messageTimestamps[msg.Header.MessageID] = report.Timestamp
+			// Check if we've exceeded encoding size limits
+			if !stop {
+				stop = totalMsgs >= lenientMaxMsgsPerObs ||
+					exceedsMaxEncodingSize(observation, p.ocrTypeCodec, lenientMaxObservationLength)
 			}
-			availableReports[srcChain] = append(availableReports[srcChain], report)
 		}
-		// Remove empty chains.
-		if len(messageObs[srcChain]) == 0 {
-			delete(messageObs, srcChain)
+
+		if stop {
+			lggr.Infow("Stop processing messages, observation is too large")
+			break
 		}
 	}
-	return messageObs, availableReports, messageTimestamps
+
+	return observation, nil
+}
+
+// initSourceChainMaps initializes maps for a source chain if they don't exist
+func initSourceChainMaps(
+	srcChain cciptypes.ChainSelector,
+	messageObs exectypes.MessageObservations,
+	msgHashes exectypes.MessageHashes,
+) {
+	if _, ok := messageObs[srcChain]; !ok {
+		messageObs[srcChain] = make(map[cciptypes.SeqNum]cciptypes.Message)
+		msgHashes[srcChain] = make(map[cciptypes.SeqNum]cciptypes.Bytes32)
+	}
+}
+
+// readMessagesForReport reads messages for the given report and validates they conform to the sequence range
+func (p *Plugin) readMessagesForReport(
+	ctx context.Context,
+	lggr logger.Logger,
+	srcChain cciptypes.ChainSelector,
+	report exectypes.CommitData,
+) ([]cciptypes.Message, error) {
+	msgs, err := p.ccipReader.MsgsBetweenSeqNums(ctx, srcChain, report.SequenceNumberRange)
+	if err != nil {
+		lggr.Errorw("unable to read all messages for report",
+			"srcChain", srcChain,
+			"seqRange", report.SequenceNumberRange,
+			"merkleRoot", report.MerkleRoot,
+			"err", err,
+		)
+		return nil, err
```

Comment:
nit: seems cleaner to return a wrapped error and log that error at the call site.